### PR TITLE
fix #144: Nested layouts infinite recursive call. 

### LINKED ISSUE #144: Upgrading from 0.7.6 to 0.7.7 breaks Nested/Cumulative Layouts.
After upgrading from `0.7.6` to `0.7.7`, layouts seem to be breaking.

image

I tried...
✅ `vike-solid@0.7.6` and `vike@0.4.210`. It works. (using `vite@5.4.11`)
❌ `vike-solid@0.7.7` and `vike@0.4.210`. It doesn't work. (using `vite@5.4.11`)
❌ `vike-solid@0.7.7` and `vike@0.4.206`. It doesn't work. (using `vite@5.4.11`)

I also thought maybe it was just Vite, but no. Because:
❌ `vike-solid@0.7.7` and `vike@0.4.210`. It doesn't work. (using `vite@6.0.3`)
✅ `vike-solid@0.7.6` and `vike@0.4.210`. It works. (using `vite@6.0.3`)

So I kinda conclude It's because of vike-solid@0.7.7.

Still investigating what the cause is... So opening this issue for that.
I'm trying to compare the changes from 0.7.6 to 0.7.7: https://github.com/vikejs/vike-solid/compare/94bda74...16c9c82 (I just used the `release: 0.7.6` and `release: 0.7.7` commits).

## Reproduction

I will add here when I can, but you can maybe use my boilerplate: https://github.com/blankeos/solid-hop (It's pretty barebones). And then reproduce the package versions I mentioned above.

### PR DESCRIPTION
**Pull request by Blankeos**
Fixes https://github.com/vikejs/vike-solid/issues/144

TL;DR; These are the two issues fixed:

- [x] Infinite Recursion (Error: Maximum call stack size exceeded)
- [x] Layout states not persisting when navigating into new [+Layout.tsx, +Layout.tsx, +Layout.tsx]

### CODE REVIEW

**Comment by Blankeos** [comment]
Have fully tested this. It works as expected. 

I wrote most of my notes as I solved the bugs in the issue comments. 

Feel free to check there.

**Comment by Blankeos** [comment]
Other solutions I attempted to fix this.

This was actually the solution that made me arrive to the simpler solution in this PR. This is longer because it rewrites the the recursion (I thought it was the culprit). 

- I initially thought it was the `createComponent` recursion that was causing the bug for non-persistence of state.
- Got this solution from Claude. This is interesting because this showcases how you can simulate "early returns" in SolidJS [(which is supposedly impossible)](https://stackoverflow.com/questions/78037797/why-cant-i-have-early-returns-in-solidjs) . Using the `<Switch>` and `<Match>` primitives.
- In the end, I noticed it was because this is reversing and accessing the array using `[i]` instead of `.at(i)`. Which actually solved the problem.

```ts
function RecursiveWrapper(props: { wrappers: FlowComponent[], children: JSX.Element, index: number }) {
 return (
   <Switch fallback={props.children}>
      <Match when={props.index < props.wrappers.length}>
        {(() => {
          // const CurrentComponent = props.wrappers.at(-(props.index + 1)) ?? Passthrough; // Access in reverse
          const CurrentComponent = props.wrappers.at(props.index) ?? Passthrough;
          
          return (
            <CurrentComponent>
              <RecursiveWrapper wrappers={props.wrappers} index={props.index + 1}>
                {props.children}
              </RecursiveWrapper>
            </CurrentComponent>
          );
        })()}
      </Match>
    </Switch>
 )
}


 function Wrapper(props: { children: JSX.Element }) {
   const pageContext = usePageContext();
 
   const [wrappers, setWrappers] = createStore<FlowComponent[]>([]);
 
   createComputed(() => {
     setWrappers(
       reconcile([
         // Inner wrapping
         ...(pageContext.config.Layout || []),
         // Outer wrapping
         ...(pageContext.config.Wrapper || []),
      ]),
        ].toReversed()),
     );
   });
  
  return (
    <RecursiveWrapper wrappers={wrappers} index={0}>    
      {props.children}
   </RecursiveWrapper>
  )
}
```

**Comment by brillout** [comment]
@magne4000 WDYT?

**Comment by magne4000** [comment]
This looks perfect, thanks!

Note: I don't know if those are limitations of SolidJS or actual bugs, we should probably look into that.

**Comment by Blankeos** [comment]
Thanks Joël! I agree, planning to make small repros of the issue those limitations with SolidJS to confirm as well.

**Comment by magne4000** [comment]
Published as `0.7.8`

**Comment by magne4000** [comment]
> Also btw I think I forgot to `pnpm format` so the CI failed.

No worries, I fixed it

**Comment by brillout** [comment]
@Blankeos I talked with Joël and, if that's something you'd be up for, we would like to add you as one of the maintainers of `vike-solid`. Let me know if that sounds appealing to you and I'll add you to the repository and the npm package.

**Comment by Blankeos** [comment]
Hi @brillout! would love to. Feel free to contact me on Discord :)
